### PR TITLE
[builder-worker] Re-work job streaming & introduce log sections.

### LINKED ISSUE #93: Stream Docker stderr to build log [was: Docker image export not working?]
@indradhanush commented on [Fri Jan 05 2018](https://github.com/habitat-sh/habitat/issues/4418)

Docker image exports are currently not working for us. We run on top of master (6e39b7a) with the Vagrant setup from [PR](https://github.com/habitat-sh/habitat/pull/4161) and the pending bug fix from [here](https://github.com/habitat-sh/habitat/pull/4392). After connecting a new plan with Docker integration, we see the docker exporter failing silently midway through building the image. We see the following logs:

**Web UI logs**

```
Sending build context to Docker daemon  158.6MB
Step 1/11 : FROM scratch
 ---> 
Step 2/11 : ENV PATH /bin
 ---> Running in 77481113cc61
 ---> 4148ca6b6ac1
Removing intermediate container 77481113cc61
Step 3/11 : ADD rootfs /
 ---> 8c0dada5a645
Step 4/11 : USER 0:0
 ---> Running in 6201a153ede1
 ---> 27a5fb8962f3
Removing intermediate container 6201a153ede1
Step 5/11 : RUN mkdir -p /hab/svc/zero-dep/data
 ---> Running in 14cd81545099
Removing intermediate container 14cd81545099
--- END: Docker export ---
```

**Worker logs**

```
worker.1    | DEBUG:habitat_builder_worker::runner: Found runnable package, running docker export
worker.1    | DEBUG:habitat_core::fs: resolved absolute path to program, program=dockerd, ident=core/docker, abs_path=/hab/pkgs/core/docker/17.09.0/20171001205930/bin/dockerd
worker.1    | DEBUG:habitat_builder_worker::runner::docker: building dockerd command, cmd="/hab/pkgs/core/docker/17.09.0/20171001205930/bin/dockerd" "-H" "unix:///hab/svc/builder-worker/data/888124392435916800/dockerd/var/run/docker.sock" "--pidfile" "/hab/svc/builder-worker/data/888124392435916800/dockerd/var/run/docker.pid" "--data-root" "/hab/svc/builder-worker/data/888124392435916800/dockerd/var/lib/docker" "--exec-root" "/hab/svc/builder-worker/data/888124392435916800/dockerd/var/run/docker" "--config-file" "/hab/svc/builder-worker/data/888124392435916800/dockerd/etc/daemon.json" "--group" "hab" "--iptables=false" "--ip-masq=false" "--ipv6=false" "--raw-logs"
worker.1    | DEBUG:habitat_builder_worker::runner::docker: setting docker export command env, PATH=/hab/pkgs/core/docker/17.09.0/20171001205930/bin
worker.1    | DEBUG:habitat_builder_worker::runner::docker: spawning dockerd export command
api.1       | DEBUG:builder_http_gateway::http::helpers: extract_pagination range: (start, end): (0, 49)
worker.1    | DEBUG:habitat_core::fs: resolved absolute path to program, program=hab-pkg-export-docker, ident=core/hab-pkg-export-docker, abs_path=/hab/pkgs/core/hab-pkg-export-docker/0.51.0/20171218213726/bin/hab-pkg-export-docker
worker.1    | DEBUG:habitat_builder_worker::runner::docker: building docker export command, cmd=building docker export command, cmd="/hab/pkgs/core/hab-pkg-export-docker/0.51.0/20171218213726/bin/hab-pkg-export-docker" "--image-name" "<username-redacted>/zero-dep" "--base-pkgs-url" "http://localhost:9636" "--url" "http://localhost:9636" "--tag-latest" "--tag-version" "--tag-version-release" "--push-image" "--username" "<username-redacted>" "--password" "<password-redacted>" "--rm-image" "--registry-type" "docker" "/hab/svc/builder-worker/data/888124392435916800/out/<username-redacted>-zero-dep-0.1.0-20180105075532-x86_64-linux.hart"
worker.1    | DEBUG:habitat_builder_worker::runner::docker: setting docker export command env, DOCKER_HOST=unix:///hab/svc/builder-worker/data/888124392435916800/dockerd/var/run/docker.sock
worker.1    | DEBUG:habitat_builder_worker::runner::docker: spawning docker export command
api.1       | DEBUG:habitat_depot::server: Headers = hab-pkg-export-docker/0.51.0/20171218213726 (x86_64-linux; 4.10.0-19-generic)
api.1       | DEBUG:habitat_depot::server: Headers = hab-pkg-export-docker/0.51.0/20171218213726 (x86_64-linux; 4.10.0-19-generic)
api.1       | DEBUG:habitat_depot::server: Headers = hab-pkg-export-docker/0.51.0/20171218213726 (x86_64-linux; 4.10.0-19-generic)
api.1       | DEBUG:habitat_depot::server: Headers = hab-pkg-export-docker/0.51.0/20171218213726 (x86_64-linux; 4.10.0-19-generic)
api.1       | DEBUG:habitat_depot::server: Headers = hab-pkg-export-docker/0.51.0/20171218213726 (x86_64-linux; 4.10.0-19-generic)
api.1       | DEBUG:habitat_depot::server: Headers = hab-pkg-export-docker/0.51.0/20171218213726 (x86_64-linux; 4.10.0-19-generic)
api.1       | DEBUG:habitat_depot::server: Headers = hab-pkg-export-docker/0.51.0/20171218213726 (x86_64-linux; 4.10.0-19-generic)
api.1       | DEBUG:habitat_depot::server: Headers = hab-pkg-export-docker/0.51.0/20171218213726 (x86_64-linux; 4.10.0-19-generic)
api.1       | DEBUG:habitat_depot::server: Headers = hab-pkg-export-docker/0.51.0/20171218213726 (x86_64-linux; 4.10.0-19-generic)
api.1       | DEBUG:habitat_depot::server: Headers = hab-pkg-export-docker/0.51.0/20171218213726 (x86_64-linux; 4.10.0-19-generic)
worker.1    | DEBUG:habitat_builder_worker::heartbeat: heartbeat pulsed: endpoint: "29614@vagrant" os: Linux state: Busy
jobsrv.1    | DEBUG:habitat_builder_jobsrv::server::worker_manager: Got heartbeat: endpoint: "29614@vagrant" os: Linux state: Busy
worker.1    | DEBUG:habitat_builder_worker::heartbeat: heartbeat pulsed: endpoint: "29614@vagrant" os: Linux state: Busy
jobsrv.1    | DEBUG:habitat_builder_jobsrv::server::worker_manager: Got heartbeat: endpoint: "29614@vagrant" os: Linux state: Busy
jobsrv.1    | DEBUG:habitat_builder_jobsrv::server::scheduler: Process status: found 0 updated jobs
worker.1    | DEBUG:habitat_builder_worker::runner::docker: completed docker export command, status=ExitStatus(ExitStatus(256))
worker.1    | DEBUG:habitat_builder_worker::runner::docker: signaling dockerd to shutdown pid=30869, sig=TERM
worker.1    | DEBUG:habitat_builder_worker::runner::docker: terminated dockerd
worker.1    | DEBUG:habitat_builder_worker::runner::log_pipe: completing log_forwarder, job_id=888124392435916800
jobsrv.1    | DEBUG:habitat_builder_jobsrv::server::log_ingester: Log complete for job 888124392435916800
jobsrv.1    | DEBUG:habitat_builder_jobsrv::server::log_ingester: Archived log for job 888124392435916800
worker.1    | DEBUG:habitat_builder_worker::runner: Failed studio build for indradhanush/zero-dep, err=BuildFailure(1)
worker.1    | WARN:habitat_builder_worker::runner: Failed to remove studio dir /hab/svc/builder-worker/data/888124392435916800/studio, err: Error { repr: Os { code: 2, message: "No such file or directory" } }
```

**dockerd logs**

```
time="2018-01-05T09:06:09.783395854Z" level=info msg="libcontainerd: new containerd process, pid: 1587"
time="2018-01-05T09:06:09.805879171Z" level=warning msg="containerd: low RLIMIT_NOFILE changing to max" current=1024 max=1048576
time="2018-01-05T09:06:10.790901720Z" level=warning msg="failed to rename /hab/svc/builder-worker/data/888159704012914688/dockerd/var/lib/docker/tmp for background deletion: rename /hab/svc/builder-worker/data/888159704012914688/dockerd/var/lib/docker/tmp /hab/svc/builder-worker/data/888159704012914688/dockerd/var/lib/docker/tmp-old: no such file or directory. Deleting synchronously"
time="2018-01-05T09:06:10.793746824Z" level=error msg="AppArmor enabled on system but the docker-default profile could not be loaded: running `apparmor_parser apparmor_parser --version` failed with output:
error: exec: \"apparmor_parser\": executable file not found in $PATH"
time="2018-01-05T09:06:12.024688467Z" level=info msg="Graph migration to content-addressability took 0.00 seconds"
time="2018-01-05T09:06:12.025451076Z" level=warning msg="Your kernel does not support swap memory limit"
time="2018-01-05T09:06:12.025565211Z" level=warning msg="Your kernel does not support cgroup rt period"
time="2018-01-05T09:06:12.025602010Z" level=warning msg="Your kernel does not support cgroup rt runtime"
time="2018-01-05T09:06:12.026884773Z" level=info msg="Loading containers: start."
time="2018-01-05T09:06:12.086503769Z" level=warning msg="Running modprobe nf_nat failed with message: ``, error: exec: \"modprobe\": executable file not found in $PATH"
time="2018-01-05T09:06:12.086686628Z" level=warning msg="Running modprobe xt_conntrack failed with message: ``, error: exec: \"modprobe\": executable file not found in $PATH"
time="2018-01-05T09:06:12.095347526Z" level=warning msg="Failed to create DOCKER-USER chain: Iptables not found"
time="2018-01-05T09:06:12.108064234Z" level=warning msg="Failed to create DOCKER-USER chain: Iptables not found"
time="2018-01-05T09:06:12.109124276Z" level=info msg="Default bridge (docker0) is assigned with an IP address 172.17.0.0/16. Daemon option --bip can be used to set a preferred IP address"
time="2018-01-05T09:06:12.121900452Z" level=warning msg="Failed to create DOCKER-USER chain: Iptables not found"
time="2018-01-05T09:06:12.121960134Z" level=info msg="Loading containers: done."
time="2018-01-05T09:06:12.176709096Z" level=info msg="Docker daemon" commit=afdb6d4 graphdriver(s)=overlay2 version=17.09.0-ce
time="2018-01-05T09:06:12.177189834Z" level=info msg="Daemon has completed initialization"
time="2018-01-05T09:06:12.217401295Z" level=info msg="API listen on /hab/svc/builder-worker/data/888159704012914688/dockerd/var/run/docker.sock"
time="2018-01-05T09:06:28.363454302Z" level=info msg="Layer sha256:5f70bf18a086007016e948b04aed3b82103a36bea41755b6cddfaf10ace3c6ef cleaned up"
time="2018-01-05T09:06:34.837588807Z" level=info msg="Layer sha256:ec3b88c0faddf0baefd51243f96b92697aa95b4f03aad930f6ea08dbde9046af cleaned up"
time="2018-01-05T09:06:35.692015969Z" level=info msg="Processing signal 'terminated'"
time="2018-01-05T09:06:35.74920319Z" level=info msg="stopping containerd after receiving terminated"
```

What could be missing here?
  

---

@eeyun commented on [Tue Jan 16 2018](https://github.com/habitat-sh/habitat/issues/4418#issuecomment-358059368)

We're going to sort this into accepted minor and try to validate/repro in our acceptance environment! Thank you for submitting!

### PR DESCRIPTION
**Pull request by fnichol**
This change consists of 2 changes which are lightly related but more
coupled in implementation (hence one commit).

Re-implement Job Streaming In `JobStreamer`
-------------------------------------------

A new implementation of job streaming was written to replace the
`LogPipe` struct, which is called a `JobStreamer`. Under the covers it
uses the same zeromq-based code but its stream handling and API is
different enough to warrent a new name.

A `JobStreamer` contains a clonable `StreamTarget` which is managed via
a `Mutex` and is the struct that contains the socket and line count
state for the stream connection. The `Arc`-clonable type of the target
allows it to be shared among multiple threads which may be concurrently
attempting to write log lines to the single log stream. As the log
stream is not read from, but rather written to, a `Mutex` seemed to be
the natural locking mechanism. Additional care was taken to prevent a
mutex deadlock by keeping calls to the stream target short and scoped
small.

The primary consuming API is a `consume_child()` function which takes a
mutable reference to a `Child` process. This method will consume both
the process' `stdout` and `stderr` streams and attempts to keep the
output ordering as close to original as possible. Additionally, there is
a `println_stderr()` function which can be used to print error messaging
to the log stream if a particular part of the build goes wrong (i.e.
during a repo clone, downloading origin keys, etc.).

Introduce Log Sections
----------------------

In addition to the `JobStreamer` work, this change introduces a more
formal notion of sections within a log stream. Future work may introduce
new sections, remove others, or possibly change the section delimiting
format, but this is a first step with 2 goals in mind:

1. Reduce the visual importance of the `--- BEGIN: Studio build ---`
lines that were starting to overwhelm the log output
2. Introduce explict start and ends to every section started into,
whether or not the section resulted in a failure. Additionally an
explicit start and end of the log was added to allow a user to know when
the log was indeed finished.

The section names are user-facing one-word tokens which should help
convey what is going on without being too vague. This is why the
previous `"POSTPROCESSING"` part of the log is now a `"publish_package"`
section name.

Closes #93

Signed-off-by: Fletcher Nichol <fnichol@nichol.ca>

### CODE REVIEW

**Comment by thesentinels** [comment]
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

**Comment by fnichol** [comment]
Some user-facing previews:

log-top

Note that all the `DEBUG:` lines in the above image were from a development instance of Builder with debugging turned on. Prior to this change, none of these lines would have been output, and even if they were, they would all be grouped at the very bottom of the log which isn't helpful when trying to associate debugging/trace output with failures.

log-bottom

**Comment by chefsalim** [comment]
tenor-8110403
tenor-38976243

**Comment by christophermaier** [comment]
tenor-54932080

**Comment by raskchanky** [comment]
tenor-95435495

**Comment by fnichol** [comment]
@thesentinels approve

File: components/builder-worker/src/runner/job_streamer.rs [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+    /// log stream on the socket.
+    ///
+    /// # Panics
+    ///
+    /// * If the protobuf struct cannot be serialized into bytes
+    ///
+    /// # Errors
+    ///
+    /// * If a message couldn't be sent successfully to the stream target socket
+    fn stream_line<S: Into<String>>(&mut self, id: u64, line: S) -> Result<()> {
```
Comment (baumanj) [inline components/builder-worker/src/runner/job_streamer.rs]: Alternatively, could this be
```rust
fn stream_line(&mut self, id: u64, line: &str) -> Result<()> {
    …
    let mut line: String = line.into();
    // or
    let mut line = line.to_string();
```
?

Eliminating the `<S: Into<String>>` makes things a bit simpler, but I wasn't sure if that provided something I didn't understand.

**Pull request merged**